File: frontend/src/schemas/scale-plan.schema.ts
import { z } from 'zod'

import {
  OrderTypeEnum,
  ScalePlanStatusEnum,
  ScalePlanTypeEnum,
  ScaleTradeTypeEnum,
} from '@/enums/scale-plan.enum'
import { ExecutionSchema } from '@/schemas/execution.schema.ts'

export const ScalePlanSchema = z.object({
  id: z.string().uuid(),
  tradeId: z.string().uuid(),
  orderType: OrderTypeEnum,
  label: z.string().min(1, 'Label is required'),
  status: ScalePlanStatusEnum,
  planType: ScalePlanTypeEnum,
  targetPrice: z.number().optional(),
  qty: z.number().min(1, 'Qty is required'),
  notes: z.string().optional(),
  goodTillDate: z.date().optional(),
  stopPrice: z.number().optional(),
  limitPrice: z.number().optional(),
  executions: z.array(ExecutionSchema),
  tradeType: ScaleTradeTypeEnum,
})

export const ScalePlanCreateSchema = ScalePlanSchema.omit({
  id: true,
  status: true,
  tradeId: true,
  executions: true,
})

export const ScalePlanUpdateSchema = ScalePlanSchema.omit({
  id: true,
  tradeId: true,
  executions: true,
}).partial()

File: backend/domain/scale_plan/scale_plan_schema.py
from datetime import datetime
from typing import Optional
from pydantic import Field

from core.base_schema import BaseSchema
from database.models import ScalePlanStatus, OrderType, PlanType, TradeType
from domain.execution.execution_schema import ExecutionRead


class ScalePlanBase(BaseSchema):
    order_type: OrderType
    label: str
    qty: float
    target_price: Optional[float] = None
    notes: str
    good_till: Optional[datetime] = None
    stop_price: Optional[float] = None
    limit_price: Optional[float] = None
    plan_type: PlanType
    trade_type: TradeType


class ScalePlanCreate(ScalePlanBase):
    pass


class ScalePlanCreateWithTradeId(ScalePlanBase):
    trade_id: str


class ScalePlanRead(ScalePlanBase):
    id: str
    trade_id: str
    trade_type: TradeType
    status: ScalePlanStatus
    executions: list[ExecutionRead] = Field(default_factory=list)


class ScalePlanCreateResponse(ScalePlanBase):
    id: str
    trade_id: str
    status: ScalePlanStatus


class ScalePlanUpdate(ScalePlanBase):
    plan_type: Optional[PlanType] = None
    order_type: Optional[OrderType] = None
    label: Optional[str] = None
    qty: Optional[float] = None
    target_price: Optional[float] = None
    notes: Optional[str] = None
    good_till: Optional[datetime] = None
    stop_price: Optional[float] = None
    limit_price: Optional[float] = None
    status: Optional[ScalePlanStatus] = None
    trade_type: Optional[TradeType] = None
